go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+			"Regex capture did not match: %v", re)
+	}
+	state.ActiveString = match[1]
+	return state, nil
+}
+
+func pvlStepFetch(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	fetchType, err := ins.AtKey(PvlFetch).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get fetch type")
+	}
+	if state.FetchResult != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Script cannot contain more than one fetch")
+	}
+	if state.Service == keybase1.ProofType_DNS {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Script cannot fetch for DNS")
+	}
+
+	switch PvlMode(fetchType) {
+	case PvlModeString:
+		res, err := g.GetExternalAPI().GetText(NewAPIArg(state.FetchURL))
+		if err != nil {
+			return state, XapiError(err, state.FetchURL)
+		}
+		state.FetchResult = &PvlFetchResult{
+			Mode:   PvlModeString,
+			String: res.Body,
+		}
+		state.ActiveString = state.FetchResult.String
+		return state, nil
+	case PvlModeJSON:
+		res, err := g.GetExternalAPI().Get(NewAPIArg(state.FetchURL))
+		if err != nil {
+			return state, XapiError(err, state.FetchURL)
+		}
+		state.FetchResult = &PvlFetchResult{
+			Mode: PvlModeJSON,
+			JSON: res.Body,
+		}
+		state.ActiveString = ""
+		return state, nil
+	case PvlModeHTML:
+		res, err := g.GetExternalAPI().GetHTML(NewAPIArg(state.FetchURL))
+		if err != nil {
+			return state, XapiError(err, state.FetchURL)
+		}
+		state.FetchResult = &PvlFetchResult{
+			Mode: PvlModeHTML,
+			HTML: res.GoQuery,
+		}
+		state.ActiveString = ""
+		return state, nil
+	default:
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Unsupported fetch type %v", fetchType)
+	}
+}
+
+func pvlStepSelectorJSON(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	if state.FetchResult == nil || state.FetchResult.Mode != PvlModeJSON {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Cannot use json selector with non-json fetch result")
+	}
+
+	selectorsw, err := ins.AtKey(PvlSelectorJSON).ToArray()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Cannot use css selector with non-html fetch result")
+	}
+
+	selectors, err := pvlJSONUnpackArray(selectorsw)
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not unpack json selector list: %v", err)
+	}
+	if len(selectors) < 1 {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Json selector list must contain at least 1 element")
+	}
+
+	results, perr := pvlRunSelectorJSONInner(g, state.FetchResult.JSON, selectors)
+	if perr != nil {
+		return state, perr
+	}
+	if len(results) < 1 {
+		return state, NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Json selector did not match any values")
+	}
+	s := strings.Join(results, " ")
+
+	state.ActiveString = s
```

Comment:
One final thought from me about `ActiveString`. Miles and I have already talked about it, nothing new here, just writing it out:

    { "selector_json": [0, "data", "children", 0, "data", "author"] },
    { "assert_regex_match": "/^%{username_service}$/i" },

These lines from the reddit PVL above are implicitly operating on `ActiveString`. An alternative to doing that would be to have every operation (both loads/selectors, and asserts) operate on a named string. The simplest version of this idea might be disappointingly verbose:

    { "selector_json": [0, "data", "children", 0, "data", "author"], "name": "author_name" },
    { "assert_regex_match": "/^%{username_service}$/i", "name": "author_name" },

But if we pair it with some other cleanup, it's possible the result could end up both easier to read and easier to validate.

    [ "selector_json",      "author_name", [0, "data", "children", 0, "data", "author"] ],
    [ "assert_regex_match", "author_name", "/^%{username_service}$/i" ],

Not sure how this would apply to `regex_capture`. Currently that one reads the active string (unlike `selector_*`, which read the full fetched document), and also writes to the active string. A different version that reads and writes to two named variables would definitely be more verbose, but possibly more flexible and less confusing? If we ever decided we wanted to have proofs that fetch *two* documents, and compare strings *between* them, some kind of naming will probably be necessary. (Also gives you slightly better error messages for free.)

Anyway, end of brain dump.